### PR TITLE
Add a send_async method to the Signal

### PR DESCRIPTION
**Pull request by pgjones**
This allows for signals to send to coroutine receivers by awaiting
them. The _async_wrapper and _sync_wrapper allows for conversion of
sync and async receivers as required if defined. If not defined a
runtime error is raised.

The wrappers are used to avoid any direct tie into asyncio, trio,
greenbacks, asgiref, or other specific async implementation.

### CODE REVIEW

**Comment by pgjones** [comment]
Note that if async coroutines are connected to any signal's `receiver_connected` or `receiver_disconnected` there will be a runtime error. I think this is acceptable.

**Comment by DonDebonair** [comment]
I saw that previous code to supprt async was reverted. Can we merge this one in and get a release soon?

**Comment by davidism** [comment]
We'll merge this when we're ready and have time.

**Comment by leandrorebelo** [comment]
Hey @davidism, is there some news about the MR?

File: blinker/base.py [hunk 1]
```diff
@@ -231,17 +232,53 @@ def temporarily_connected_to(self, receiver, sender=ANY):
...
+          with ``None``. Only accepts one positional argument.
+        :param _sync_wrapper: A callable that should wrap a synchronous
+          receiver and run it when awaited.
+
+        :param kwargs: Data to be sent to receivers.
+        """
+        sender = self._extract_sender(sender)
+        results = []
+        for receiver in self.receivers_for(sender):
+            if is_coroutine_function(receiver):
```
Comment (AndyFug) [inline blinker/base.py:274]: Should this line (274) be the inverse logic?
`if not is_coroutine_function(receiver)`
Certainly this small tweak made it work for me... apologies if I've misunderstood the concept.

Comment (pgjones) [reply blinker/base.py:274]: Yep, thanks

**Comment by pgjones** [comment]
@Secrus what do you make of this?

File: requirements/tests.txt [hunk 2]
```diff
@@ -7,23 +7,15 @@
...
     # via pytest
 pluggy==1.0.0
     # via pytest
 pytest==7.2.1
+    # via
+    #   -r requirements/tests.in
+    #   pytest-asyncio
+pytest-asyncio==0.20.3
     # via -r requirements/tests.in
-tomli==2.0.1
```
Comment (Secrus) [inline requirements/tests.txt:24]: Please relock on python 3.7 so all the dependencies of `pytest` are locked.

File: src/blinker/_utilities.py [hunk 3]
```diff
@@ -93,3 +98,35 @@ def __get__(self, obj, cls):
         value = self._deferred(obj)
         setattr(obj, self._deferred.__name__, value)
         return value
+
+
+def is_coroutine_function(func: Any) -> bool:
```
Comment (Secrus) [inline src/blinker/_utilities.py:103]: Please add TODO comment, scheduling this to be removed with Python 3.7 support

File: src/blinker/base.py [hunk 4]
```diff
@@ -228,17 +229,53 @@ def temporarily_connected_to(self, receiver, sender=ANY):
...
+        results = []
+        for receiver in self.receivers_for(sender):
+            if is_coroutine_function(receiver):
+                if _async_wrapper is None:
+                    raise RuntimeError("Cannot send to a coroutine function")
+                receiver = _async_wrapper(receiver)
+            results.append((receiver, receiver(sender, **kwargs)))
+        return results
+
+    async def send_async(self, *sender, _sync_wrapper=None, **kwargs):
```
Comment (Secrus) [inline src/blinker/base.py:255]: I am not used to working with `async`, so let me check if I understand what is happening here. This allows connecting `async` functions to signals, and with proper `_sync_wrapper`, sync functions will also work, and the other way around in `send`?

**Review by Secrus — CHANGES_REQUESTED** [review]
This surely needs a proper documentation entry.

File: src/blinker/_utilities.py (hunk 3 above)
Comment (pgjones) [reply src/blinker/_utilities.py:103]: I'm not sure it is a TODO, as nothing can be done at the moment. (I tend to see TODOs as indicating something that would have been done if time allowed). This said, I don't mind so if you still think a TODO should be added please say so.

File: src/blinker/base.py (hunk 4 above)
Comment (pgjones) [reply src/blinker/base.py:255]: Yep

File: requirements/tests.txt (hunk 2 above)
Comment (pgjones) [reply requirements/tests.txt:24]: Done

**Comment by pgjones** [comment]
I've added documentation

File: src/blinker/_utilities.py (hunk 3 above)
Comment (Secrus) [reply src/blinker/_utilities.py:103]: ok, it's no big deal, let's leave it as it is

**Comment by djsilcock** [comment]
Just as an observation on the principle of least surprise, I would have intuitively expected `await some_signal.send_async()` to run the handlers concurrently, rather than sequentially - the concurrent behaviour would require something like:

    res=[(rec,await result if asyncio.isfuture(result) else result) for rec,result in some_signal.send(_async_wrapper=asyncio.create_task)

**Pull request merged**